Detected Oracle driver stuck during Statement initialization Caused by: java.sql.SQLRecoverableException: Closed Connection

I have local configuration: oracle container (19.3) + maven container and it works fine.
Here is output:

> 
> INFO] --- utplsql-maven-plugin:3.1.4:test (run-plsql-test-packages) @ debtcollect ---
> [INFO] utPLSQL Version = v3.1.10.3349
> [INFO] Invoking TestRunner with: /opt/debtcollect/target
> [INFO] TestRunner initialized
> [INFO] Running on utPLSQL v3.1.10.3349
> [INFO] Initializing reporters
> [INFO] Running tests
> [INFO] Running tests finished.
> [info] Writing report UT_COVERAGE_SONAR_REPORTER to /opt/debtcollect/target/coverage-sonar-report.xml
> [info] Writing report UT_COVERAGE_HTML_REPORTER to /opt/debtcollect/target/coverage-report.html
> [info] Writing report UT_SONAR_TEST_REPORTER to /opt/debtcollect/target/sonar-test-report.xml
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  4.462 s
> [INFO] Finished at: 2022-04-17T15:38:41Z
> [INFO] ------------------------------------------------------------------------

And I have test oracle instance in my company (19.6) + my local maven and i've got strange error with connection.
Here is output:
> [INFO] --- utplsql-maven-plugin:3.1.4:test (run-plsql-test-packages) @ debtcollect ---
> [INFO] utPLSQL Version = v3.1.10.3349
> [INFO] Invoking TestRunner with: /Users/ivan/Documents/Work/sandbox/debt/target
> [INFO] TestRunner initialized
> [INFO] Running on utPLSQL v3.1.10.3349
> [INFO] Initializing reporters
> [ERROR] Detected Oracle driver stuck during Statement initialization
> [ERROR] Unexpected error opening file ouput 
> org.apache.maven.plugin.MojoExecutionException: Unexpected error opening file ouput 
>     at org.utplsql.maven.plugin.reporter.ReporterWriter.writeReports (ReporterWriter.java:101)
>     at org.utplsql.maven.plugin.reporter.ReporterWriter.writeReporters (ReporterWriter.java:63)
>     at org.utplsql.maven.plugin.UtPLSQLMojo.execute (UtPLSQLMojo.java:196)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
>     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke (Method.java:568)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
> Caused by: java.sql.SQLRecoverableException: Closed Connection
>     at oracle.jdbc.driver.PhysicalConnection.prepareCall (PhysicalConnection.java:1935)
>     at oracle.jdbc.driver.PhysicalConnection.prepareCall (PhysicalConnection.java:1911)
>     at oracle.jdbc.driver.PhysicalConnection.prepareCall (PhysicalConnection.java:1877)
>     at org.utplsql.api.outputBuffer.OutputBufferProvider.hasOutput (OutputBufferProvider.java:62)
>     at org.utplsql.api.outputBuffer.OutputBufferProvider.getCompatibleOutputBuffer (OutputBufferProvider.java:33)
>     at org.utplsql.maven.plugin.reporter.ReporterWriter.writeReports (ReporterWriter.java:74)
>     at org.utplsql.maven.plugin.reporter.ReporterWriter.writeReporters (ReporterWriter.java:63)
>     at org.utplsql.maven.plugin.UtPLSQLMojo.execute (UtPLSQLMojo.java:196)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
>     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke (Method.java:568)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  14.796 s
> [INFO] Finished at: 2022-04-17T18:35:11+03:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal org.utplsql:utplsql-maven-plugin:3.1.4:test (run-plsql-test-packages) on project debtcollect: Oracle driver stuck during creating the TestRunner statement. Retry.: TimeoutException -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

The driver is the same, the uptplsql version is the same, the oracle version is almost the same (major = 19.0, minor is little different).
With the same oracle driver and mvn liquibase plugin i create tables and procedures on both servers, i've got this error only when try to run tests.
Oracle JDBC version is 19.3.0.0, also tried 21.3.0.0 but no changes.
Command ut.run on server works well of course on both servers.
If it's important: connection to company server is over VPN.
If it's important2: server is far away and first connection is kinda slow so local server connection is 100ms, remote server connection is 5-7 seconds.

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -39,10 +39,10 @@ dependencies {
 
     // This dependency is used internally, and not exposed to consumers on their own compile classpath.
     implementation("org.slf4j:slf4j-api:1.7.26")
-    implementation("com.oracle.ojdbc:ojdbc8:$ojdbcVersion") {
-        exclude(group = "com.oracle.ojdbc")
+    implementation("com.oracle.database.jdbc:ojdbc8:$ojdbcVersion") {
+        exclude(group = "com.oracle.database.jdbc", module = "ucp")
     }
-    implementation("com.oracle.ojdbc:orai18n:$ojdbcVersion")
+    implementation("com.oracle.database.nls:orai18n:$ojdbcVersion")
 
     // Use Jupiter test framework
     testImplementation("org.junit.jupiter:junit-jupiter:$junitVersion")

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -38,7 +38,7 @@ dependencies {
     api("com.google.code.findbugs:jsr305:3.0.2")
 
     // This dependency is used internally, and not exposed to consumers on their own compile classpath.
-    implementation("org.slf4j:slf4j-api:1.7.26")
+    implementation("org.slf4j:slf4j-api:1.7.36")
     implementation("com.oracle.database.jdbc:ojdbc8:$ojdbcVersion") {
         exclude(group = "com.oracle.database.jdbc", module = "ucp")
     }

--- a/src/main/java/org/utplsql/api/TestRunner.java
+++ b/src/main/java/org/utplsql/api/TestRunner.java
@@ -145,8 +145,8 @@ public class TestRunner {
         return this;
     }
 
-    public TestRunner catchOraStuck( boolean catchOraStuck ) {
-        this.options.catchOraStuck = catchOraStuck;
+    public TestRunner oraStuckTimeout(Integer oraStuckTimeout ) {
+        this.options.oraStuckTimeout = oraStuckTimeout;
         return this;
     }
 
@@ -218,7 +218,7 @@ public class TestRunner {
 
         TestRunnerStatement testRunnerStatement = null;
         try {
-            testRunnerStatement = ( options.catchOraStuck ) ? initStatementWithTimeout(conn) : initStatement(conn);
+            testRunnerStatement = ( options.oraStuckTimeout > 0 ) ? initStatementWithTimeout(conn, options.oraStuckTimeout) : initStatement(conn);
             logger.info("Running tests");
             testRunnerStatement.execute();
             logger.info("Running tests finished.");
@@ -236,7 +236,7 @@ public class TestRunner {
         return compatibilityProxy.getTestRunnerStatement(options, conn);
     }
 
-    private TestRunnerStatement initStatementWithTimeout( Connection conn ) throws OracleCreateStatmenetStuckException, SQLException {
+    private TestRunnerStatement initStatementWithTimeout( Connection conn, int timeout ) throws OracleCreateStatmenetStuckException, SQLException {
         ExecutorService executor = Executors.newSingleThreadExecutor();
         Callable<TestRunnerStatement> callable = () -> compatibilityProxy.getTestRunnerStatement(options, conn);
         Future<TestRunnerStatement> future = executor.submit(callable);
@@ -244,7 +244,7 @@ public class TestRunner {
         // We want to leave the statement open in case of stuck scenario
         TestRunnerStatement testRunnerStatement = null;
         try {
-            testRunnerStatement = future.get(2, TimeUnit.SECONDS);
+            testRunnerStatement = future.get(timeout, TimeUnit.SECONDS);
         } catch (TimeoutException e) {
             logger.error("Detected Oracle driver stuck during Statement initialization");
             executor.shutdownNow();

--- a/src/main/java/org/utplsql/api/TestRunnerOptions.java
+++ b/src/main/java/org/utplsql/api/TestRunnerOptions.java
@@ -30,7 +30,7 @@ public class TestRunnerOptions {
     public boolean randomTestOrder = false;
     public Integer randomTestOrderSeed;
     public final Set<String> tags = new LinkedHashSet<>();
-    public boolean catchOraStuck = false;
+    public Integer oraStuckTimeout = 0;
 
     public String getTagsAsString() {
         return String.join(",", tags);

--- a/src/main/java/org/utplsql/api/Version.java
+++ b/src/main/java/org/utplsql/api/Version.java
@@ -33,10 +33,14 @@ public class Version implements Comparable<Version> {
     public final static Version V3_1_6 = new Version("3.1.6", 3, 1, 6, 2729, true);
     public final static Version V3_1_7 = new Version("3.1.7", 3, 1, 7, 3085, true);
     public final static Version V3_1_8 = new Version("3.1.8", 3, 1, 8, 3188, true);
+    public final static Version V3_1_9 = new Version("3.1.9", 3, 1, 9, 3268, true);
+    public final static Version V3_1_10 = new Version("3.1.10", 3, 1, 10, 3347, true);
+    public final static Version V3_1_11 = new Version("3.1.11", 3, 1, 11, 3557, true);
+    public final static Version V3_1_12 = new Version("3.1.12", 3, 1, 12, 3876, true);
     private final static Map<String, Version> knownVersions =
-            Stream.of(V3_0_0, V3_0_1, V3_0_2, V3_0_3, V3_0_4, V3_1_0, V3_1_1, V3_1_2, V3_1_3, V3_1_4, V3_1_5, V3_1_6, V3_1_7, V3_1_8)
+            Stream.of(V3_0_0, V3_0_1, V3_0_2, V3_0_3, V3_0_4, V3_1_0, V3_1_1, V3_1_2, V3_1_3, V3_1_4, V3_1_5, V3_1_6, V3_1_7, V3_1_8, V3_1_9, V3_1_10, V3_1_11, V3_1_12)
                     .collect(toMap(Version::toString, Function.identity()));
-    public final static Version LATEST = V3_1_8;
+    public final static Version LATEST = V3_1_12;
 
     private final String origString;
     private final Integer major;

